src/AppInstallerRepositoryCore/CompositeSource.cpp

```diff
@@ -22,44 +22,84 @@ namespace AppInstaller::Repository
             };
         }
 
-        Pinning::PinKey GetPinKey(IPackage* availablePackage)
+        Pinning::PinKey GetPinKeyForAvailable(IPackage* availablePackage)
         {
             return {
                 availablePackage->GetProperty(PackageProperty::Id).get(),
                 availablePackage->GetLatestAvailableVersion(PinBehavior::IgnorePins)->GetSource().GetIdentifier()
             };
         }
 
-        std::optional<PackageVersionKey> GetLatestAvailableVersionKeySatisfyingPin(const std::vector<PackageVersionKey>& availableVersionKeys, PinBehavior pinBehavior)
+        // Gets the pinned state for an available PackageVersionKey that may have a pin,
+        // and optionally an additional pin that could come from the installed version.
+        // If there are both installed and available pins, we return the one that is the most strict.
+        Pinning::PinType GetPinnedStateForVersion(
+            const PackageVersionKey& availableVersionKey,
+            const std::optional<Pinning::Pin>& pin,
+            PinBehavior pinBehavior)
+        {
+            if (pinBehavior == PinBehavior::IgnorePins)
+            {
+                // No need to check anything
+                return Pinning::PinType::Unknown;
+            }
+
+            // For the pin in the available version, we can ignore it depending on the behavior and type.
+            // If it is gating, we don't need to check the version as that was already done.

```

Comment:
Do you feel we should centralize the logic to do the gated version check here so this will be the one and only logic of "what is the pinned state of the version" and everywhere else should just call this function?

Reply:
I think that's what I did? I had some duplication before but I thought I had moved it all here. Can't find another use of the `GetGatedVersion()`